File: BarcodeKanojo-iOS/BarcodeKanojo/Models/Product.swift
import Foundation

/// Product/barcode information model.
/// Source: Product.java + ProductParser.java
struct Product: Codable, Hashable, Sendable {

    var barcode: String?
    var name: String?
    var categoryId: Int
    var category: String?
    var comment: String?
    var location: String?
    /// Geo coordinates as comma-separated "lat,lng" string.
    var geo: String?
    var scanCount: Int
    var companyName: String?
    var country: String?
    var price: String?
    var product: String?

    // MARK: - Computed Properties

    var productImageURL: String? {
        guard let barcode, let product else { return nil }
        return "/product_images/barcode/\(barcode)/\(product).png"
    }

    // MARK: - Default Initializer

    init(
        barcode: String? = nil,
        name: String? = nil,
        categoryId: Int = 1,
        category: String? = nil,
        comment: String? = nil,
        location: String? = nil,
        geo: String? = nil,
        scanCount: Int = 0,
        companyName: String? = nil,
        country: String? = nil,
        price: String? = nil,
        product: String? = nil
    ) {
        self.barcode = barcode
        self.name = name
        self.categoryId = categoryId
        self.category = category
        self.comment = comment
        self.location = location
        self.geo = geo
        self.scanCount = scanCount
        self.companyName = companyName
        self.country = country
        self.price = price
        self.product = product
    }

    // MARK: - CodingKeys

    enum CodingKeys: String, CodingKey {
        case barcode, name, category, comment, location, geo, country, price, product
        case categoryId = "category_id"
        case scanCount = "scan_count"
        case companyName = "company_name"
    }

    // MARK: - Custom Decoding

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        barcode = try container.decodeIfPresent(String.self, forKey: .barcode)
        name = try container.decodeIfPresent(String.self, forKey: .name)
        categoryId = try container.decodeIfPresent(Int.self, forKey: .categoryId) ?? 1
        category = try container.decodeIfPresent(String.self, forKey: .category)
        comment = try container.decodeIfPresent(String.self, forKey: .comment)
        location = try container.decodeIfPresent(String.self, forKey: .location)
        // geo can be a "lat,lng" string or a {"lat":...,"lng":...} object
        if let geoString = try? container.decodeIfPresent(String.self, forKey: .geo) {
            geo = geoString
        } else if let geoObject = try? container.decodeIfPresent(GeoObject.self, forKey: .geo) {
            geo = "\(geoObject.lat),\(geoObject.lng)"
        } else {
            geo = nil
        }
        scanCount = try container.decodeIfPresent(Int.self, forKey: .scanCount) ?? 0
        companyName = try container.decodeIfPresent(String.self, forKey: .companyName)
        country = try container.decodeIfPresent(String.self, forKey: .country)
        price = try container.decodeIfPresent(String.self, forKey: .price)
        product = try container.decodeIfPresent(String.self, forKey: .product)
    }
}

// MARK: - Geo Object Helper

private struct GeoObject: Codable {
    let lat: Double
    let lng: Double
}

File: barcode-kanojo-server/models/product.py
from datetime import datetime

from sqlalchemy import Integer, String, Float, DateTime
from sqlalchemy.orm import Mapped, mapped_column

from models.database import Base


class Product(Base):
    __tablename__ = "products"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    barcode: Mapped[str] = mapped_column(String(50), unique=True, nullable=False)
    name: Mapped[str | None] = mapped_column(String(255))
    company_name: Mapped[str | None] = mapped_column(String(255))
    category_id: Mapped[int] = mapped_column(Integer, default=1)
    category: Mapped[str | None] = mapped_column(String(100))
    comment: Mapped[str | None] = mapped_column(String(1000))
    geo_lat: Mapped[float] = mapped_column(Float, default=0.0)
    geo_lng: Mapped[float] = mapped_column(Float, default=0.0)
    location: Mapped[str | None] = mapped_column(String(255))
    country: Mapped[str | None] = mapped_column(String(100))
    price: Mapped[str | None] = mapped_column(String(50))
    product_image_url: Mapped[str | None] = mapped_column(String(255))
    scan_count: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    def to_api_dict(self) -> dict:
        return {
            "id": self.id,
            "barcode": self.barcode,
            "name": self.name or "",
            "company_name": self.company_name or "",
            "category_id": self.category_id,
            "category": self.category or "",
            "comment": self.comment or "",
            "geo": {"lat": self.geo_lat, "lng": self.geo_lng},
            "location": self.location or "",
            "country": self.country or "",
            "price": self.price or "",
            "product_image_url": self.product_image_url or "",
            "scan_count": self.scan_count,
        }


class ProductCategory(Base):
    __tablename__ = "product_categories"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(100), nullable=False)

    def to_api_dict(self) -> dict:
        return {"id": self.id, "name": self.name}
